modules/doobie/src/main/scala/doobiemapping.scala

```diff
@@ -298,21 +309,35 @@ case class Row(elems: List[Any]) {
 }
 
 object Row {
-  def mkRead(cols: List[ColumnRef]): Read[Row] = {
-    def typeToGet(tpe: Type): Get[_] = tpe match {
-      case IntType => Get[String]
-      case FloatType => Get[Double]
-      case StringType => Get[String]
-      case BooleanType => Get[Boolean]
-      case _ => Get[String]
-    }
 
-    val gets = cols.map(col => (typeToGet(col.tpe), NoNulls))
+  // Placeholder for nulls read from non-nullable columns introduced via an outer join.
+  case object FailedJoin
 
-    def unsafeGet(rs: ResultSet, n: Int): Row = {
-      Row(gets.zipWithIndex.map { case (g, i) => g._1.unsafeGetNonNullable(rs, n+i) })
+  def mkRead(cols: List[ColumnRef], isJoin: ColumnRef => Boolean): Read[Row] = {
+
+    def typeToGet(cr: ColumnRef): (ColumnRef, (Get[_], NullabilityKnown)) =
```

Comment:
I don't understand why we're wrapping the argument in a `ColumnRef` and returning a `ColumnRef` in an outer pair. Only the `Type` component of `cr` appears to be used in the body of `typeToGet` and only `_2` of the outer pair appears to be used in the sequel (ie. only the `(Type, NullabilityKnown)` elements of `gets` is used).

Reply:
`unsafeGet` needs the `ColumnRef` to figure out whether the column is the product of an outer join or not. Is there a simpler way to figure this out?

Reply:
Right, but I'm still not quite seeing why the signature of `typeToGet` can't be,
```scala
def typeToGet(tpe: Type): (Get[_], NullabilityKnown)
```
I've reworked this area a fair bit, so it's not worth angsting over, but in case I'm missing something ...